content/blog/2019-10-09-loops/index.md

```diff
@@ -0,0 +1,235 @@
+registers, processors—in addition to how much data can sit at each point on
+the way.
+
+Today, unified memory prevents the error-prone and time consuming process of
+allocating and moving data between CPUs and GPUs [5]. This pretty much
+handles the problems posed by Carr et al. in a very limited number of
+systems. The memory is still physically separated, yet in Kepler and CUDA 6
+unified memory is currently an option for developers.
+
+#### Data Dependence
+Data dependence is an important concept within this article, so here we’ll spend a bit of time discussing it. Goff et al. put it very nicely in their paper Practical Dependence Testing:
+
+<img src="data_dependence.png" style="width: 100%">
+
+# Loop Optimizations
+#### Loop Permutation
+Loop Permutation is perhaps the most straightforward loop optimization.
+Iterating over arrays in the wrong order is one of the easiest ways to cause
+a huge amount of cache misses. It’s also the cause of an endless [number](https://stackoverflow.com/questions/33722520/why-is-iterating-2d-array-row-major-faster-than-column-major) of [SO](https://stackoverflow.com/questions/13093155/c-2d-array-access-speed-changes-based-on-ab-order)
+[questions](https://stackoverflow.com/questions/9936132/why-does-the-order-of-the-loops-affect-performance-when-iterating-over-a-2d-arra).
+
+Take these two snippets of code (can also be found online [here](http://ideone.com/PUJhdP)).
+
+A. Column-Major
+```
+int A[DIM1][DIM2];
+for (int iter = 0; iter < iters; iter++)
+    for (int j = 0; j < DIM2; j ++)
+        for (int i = 0; i < DIM1; i++)
+            A[i][j]++;
+```
+
+B. Row-Major
+```
+int A[DIM1][DIM2];
+for (int iter = 0; iter < iters; iter++)
+    for (int j = 0; j < DIM2; j ++)
+        for (int i = 0; i < DIM1; i++)
+            A[i][j]++;
+```
+
+Both of these loops perform `DIM1 * DIM2` computation. The only difference is
+whether they iterate through A in row-major or column-major order. What do
+you think their performance difference is?
+
+If you’re wise to the ways of cache locality, you might answer B. And for
+suitably large values of `DIM1` and `DIM2`, you’d be right!
+
+With DIM1=1024, DIM2=1024, and iters=1e3, we get an order of magnitude win for B!
+A (Column-Major): 4916ms
+B (Row-Major): 485ms
+
+Loop permutation captures this optimization. Intuitively, you could imagine
+the below picture when it comes to what's happening. Imagine that the grid
+represents a 2-dimensional array. However, since 2-dimensional arrays are
+actually 1-dimensional arrays in memory, the cache lines must align along a
+particular axis.
+
+Now, compare the 2 access patterns (the blue line and the red line). The blue
+access pattern loads a single cache line and then accesses all of its
+elements before moving onto the next cache line. The red access pattern,
+however, loads a single cache line, accesses a single element, and then moves
+onto the next cache line.
+
+<img src="cache_lines.jpeg" style="width: 100%">
```

Comment:
This diagram is awesome!!

Reply:
Completely thanks to @g-c-c. Their notes are also a wonder to look at.